cluster-autoscaler/simulator/dynamicresources/utils/sanitize.go

```diff
@@ -0,0 +1,166 @@
+/*
+Copyright 2024 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package utils
+
+import (
+	"fmt"
+
+	v1 "k8s.io/api/core/v1"
+	resourceapi "k8s.io/api/resource/v1beta1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/uuid"
+)
+
+// SanitizeNodeResourceSlices can be used to duplicate node-local ResourceSlices attached to a Node, when duplicating the Node. The new slices
+// are pointed to newNodeName, and nameSuffix is appended to all pool names (pool names have to be unique within a driver, so we can't
+// leave them as-is when duplicating). Returns a map of all pool names (without the suffix) that can be used with SanitizePodResourceClaims().
+// Returns an error if any of the slices isn't node-local.
+func SanitizeNodeResourceSlices(nodeLocalSlices []*resourceapi.ResourceSlice, newNodeName, nameSuffix string) (newSlices []*resourceapi.ResourceSlice, oldPoolNames map[string]bool, err error) {
+	oldPoolNames = map[string]bool{}
+	for _, slice := range nodeLocalSlices {
+		if slice.Spec.NodeName == "" {
+			return nil, nil, fmt.Errorf("can't sanitize slice %s because it isn't node-local", slice.Name)
+		}
+		sliceCopy := slice.DeepCopy()
+		sliceCopy.UID = uuid.NewUUID()
+		sliceCopy.Name = fmt.Sprintf("%s-%s", slice.Name, nameSuffix)
+		sliceCopy.Spec.Pool.Name = fmt.Sprintf("%s-%s", slice.Spec.Pool.Name, nameSuffix)
+		sliceCopy.Spec.NodeName = newNodeName
+
+		oldPoolNames[slice.Spec.Pool.Name] = true
+		newSlices = append(newSlices, sliceCopy)
+	}
+	return newSlices, oldPoolNames, nil
+}
+
+// SanitizePodResourceClaims can be used to duplicate ResourceClaims needed by a Pod, when duplicating the Pod.
+//   - ResourceClaims owned by oldOwner are duplicated and sanitized, to be owned by a duplicate pod - newOwner.
+//   - ResourceClaims not owned by oldOwner are returned unchanged in the result. They are shared claims not bound to the
+//     lifecycle of the duplicated pod, so they shouldn't be duplicated.
```

Comment:
> Does this sound correct to you?

I won't pretend that I fully understand the CA logic here, but I trust your judgment, so yes. :sweat_smile: 

> is the max length actually checked in filters?

No. Normally, a workload is set up so that it never exceeds that limit. If that happens, it gets detected by the apiserver when the scheduler plugin tries to update the ResourceClaim, so there is no need to duplicate those checks in the plugin.

In CA, you don't send anything to the apiserver for validation, so you have to replicate that in the code which tracks state.